api/internal/git/cloner.go

```diff
@@ -47,8 +67,129 @@ func ClonerUsingGitExec(repoSpec *RepoSpec) error {
+
+	permanentGitRefPath := gitRootDir.Join(repoFolderName)
+	mutableGitRefPath := filesys.ConfirmedDir(gitBranchesRootDir).Join(repoFolderName)
+
+	//search into mutable cached directory
+	if _, err := os.Stat(mutableGitRefPath); os.IsNotExist(err) {
+		//search into immutable cached dir
+		if _, err := os.Stat(permanentGitRefPath); os.IsNotExist(err) {
+			targetDir, err = cloneToCacheDir(repoSpec, mutableGitRefPath, permanentGitRefPath, acceptBranches)
+			if err != nil {
+				return err
+			}
+
+		} else { // we found it and it is an immutable
+			targetDir = filesys.ConfirmedDir(permanentGitRefPath)
+		}
+	} else {
+		if !acceptBranches {
+			return fmt.Errorf("we dont accept banch ref %s/%s", repoSpec.OrgRepo, repoSpec.Path)
+		}
+		//we found it and it is a mutable
+		targetDir = filesys.ConfirmedDir(mutableGitRefPath)
+
+		//TODO decide if we should clean old copy, update, or let the user clean the cache when he want
+	}
+
+	repoSpec.Dir = targetDir
+	return nil
+
+}
+
+func gitCacheRootDir() (filesys.ConfirmedDir, error) {
+	var cacheHome string
+	if cacheHome = os.Getenv(konfig.XdgCacheHomeEnv); cacheHome == "" {
+		usr, err := user.Current()
+		if err != nil {
+			return "", err
+		}
+		cacheHome = fmt.Sprintf("%s/%s", usr.HomeDir, konfig.XdgCacheHomeEnvDefault)
+	}
+
+	return filesys.ConfirmedDir(fmt.Sprintf("%s/%s", cacheHome, "git")), nil
+}
+
+func cloneToCacheDir(repoSpec *RepoSpec, mutableGitRefPath, permanentGitRefPath string, acceptBranches bool) (targetDir filesys.ConfirmedDir, err error) {
+	log.Printf("cloning %s", repoSpec.Raw())
+	err = ClonerUsingGitExec(repoSpec, acceptBranches)
+	targetDir = ""
+	if err != nil {
+		return
+	}
+
+	if repoSpec.IsBranchRef {
+		log.Printf("Git repo %v imported with a branch instead of tag or commitId", repoSpec.raw)
+		targetDir = filesys.ConfirmedDir(mutableGitRefPath)
+	} else {
+		targetDir = filesys.ConfirmedDir(permanentGitRefPath)
+	}
+
+	log.Printf("move %s to %s", repoSpec.Dir.String(), targetDir.String())
+	err = os.Rename(repoSpec.Dir.String(), targetDir.String())
+	if err != nil {
+		return
+	}
+	return
+}
+
+func cacheDirPrerequisites(gitRootDir filesys.ConfirmedDir, gitBranchesRootDir string) error {
+	if _, err := os.Stat(gitRootDir.String()); os.IsNotExist(err) {
+		//create permanent cache dir if it does not exist
+		err = os.Mkdir(gitRootDir.String(), 0755|os.ModeDir)
+		if err != nil {
+			return err
+		}
+	}
+	if _, err := os.Stat(gitBranchesRootDir); os.IsNotExist(err) {
+		//create mutable ref cache dir if it does not exist
+		err = os.Mkdir(gitBranchesRootDir, 0755|os.ModeDir)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// check if the git repo ref is a branch
+func isMutableRepoRef(repoSpec *RepoSpec) (bool, error) {
+	r, err := newCmdRunner(repoSpec.Timeout)
+	r.dir = repoSpec.Dir
+	if err != nil {
+		return false, err
+	}
+	origin, err := r.run("remote", "show", "origin")
```

Comment:
There should be a better command for this: you should be able to figure it out from locally known info.